COM Record field assignment using type information and multidim SAFEARRAYs as field values

geppi opened the pull request:
This PR implements:

- Respecting type information when setting COM Record field values which enables SAFEARRAY fields of various element types, not just SAFEARRY(VARIANT).
- Creating a multidimensional SAFEARRAY(COM Record) from a nested sequence of COM Records.
- Retrieving a multidimensional SAFEARRAY(COM Record) as a nested sequence of COM Records.
- Using multidimensional SAFEARRAYs of various element types as Record fields.
- Tests for the implementation.
These do also cover a test case as requested in  https://github.com/mhammond/pywin32/pull/2641#pullrequestreview-3160154481.

geppi commented:
@Avasam I had to retarget the Visual Studio project of PyCOMTest for the build tools v143 with a wild card for "the latest" installed Win SDK version because the build did fail, complaining that:

> The Windows SDK version 10.0.20348.0 was not found.

Avasam commented:
> @Avasam I had to retarget the Visual Studio project of PyCOMTest for the build tools v143 with a wild card for "the latest" installed Win SDK version because the build did fail, complaining that:

I don't mind the retargeting and bumping the platform toolset to v143 (Visual Studio 2022) makes sense. I probably missed that somewhere in the recent CI migration to `windows-2022` (and/or something else changed after the initial migration).

---

As for the rest of the PR, I am not qualified to review it, so 🤷

mhammond commented on com/TestSources/PyCOMTest/PyCOMImpl.cpp:
```diff
@@ -650,9 +650,70 @@ HRESULT CPyCOMTest::VerifyArrayOfStructs(TestStruct2 *prec, VARIANT_BOOL *is_ok)
             break;
         }
     }
+    hr = SafeArrayUnaccessData(prec->array_of_records);
```
I know this is just a test and doesn't really matter, but wouldn't returning hr on failure be better? Ditto for most other error returns below.

geppi replied:
Done

mhammond commented on com/TestSources/PyCOMTest/PyCOMTest.vcxproj:
```diff
@@ -163,4 +163,4 @@
   <ImportGroup Label="ExtensionTargets">
   </ImportGroup>
-</Project>
+</Project>
```
this can probably be reverted?

mhammond commented on com/win32com/src/oleargs.cpp:
```diff
@@ -279,18 +279,22 @@ BOOL PyCom_VariantFromPyObject(PyObject *obj, VARIANT *var)
                 return FALSE;
             V_VT(var) = VT_ARRAY | VT_RECORD;
         }
+        else if (obItemCheck != NULL && PyFloat_Check(obItemCheck)) {
```
I don't quite understand why just floats are special here and not all types? That said though, I don't fully understand why Records were always special, but at least just them being special doesn't seem as odd as arbitrarily doing the same for floats?

geppi replied:
You're right, they aren't special. Initially I also wanted to implement this for the other types.

However, I couldn't find a good solution how to do this for integers.
Using the approach taken for the non SAFEARRAY case would mean that the whole nested array would need to be scanned for the largest integer to decide which COM incarnation of an integer to use. Still, this would not necessarily satisfy the requirement defined in a type library as far as I understand it.

What I mean is the following. The type library I'm dealing with has a SAFEARRAY(double) as the field of a COM Record. That's were the existing code did fail because it only generated the SAFEARRAY(VARIANT(VT_R8)).
The type library also defines method parameters of type SAFEARRAY(long), SAFEARRAY(short) and SAFEARRAY(int). Luckily none of those are used as the field type of a Record.
The problem as far as I understand is that for method parameters the exact type that is required is known when the Python type is packaged into the VARIANTARG. Therefore there is no question if a VT_UI4, VT_I4, VT_UI8, VT_I8 or whatever other incarnation of integers is required.
This is not the case for a record field. At least I couldn't figure out from where the conversion function could get this information.

Now you can argue that also for an array of Pyfloat it is not clear if it should be converted to a SAFEARRAY(VT_R4) or SAFEARRAY(VT_R8) although the ambiguity is smaller. That's true and this is the point were I have to confess that I was simply selfish and wanted to solve my use case. ;-)

If you can help me what the best way would be to implement this, I'm happy to extend it to the other types.

I also didn't feel comfortable with the COM interface types so that would be another step were I would need help please.

mhammond commented on com/win32com/src/oleargs.cpp:
```diff
@@ -1044,20 +1045,19 @@ PyObject *PyCom_PyObjectFromSAFEARRAYBuildDimension(SAFEARRAY *psa, VARENUM vt,
+    // Get a pointer for the dimension to iterate.
+    long *pMyArrayIndex = arrayIndices + (dimNo - 1);
+    *pMyArrayIndex = lb;
+    // For Records wie arrange all BuildItems in one single RecordBuffer.
```
nit: typo wie

geppi replied:
Done

mhammond commented on com/win32com/src/PyRecord.cpp:
```diff
@@ -53,25 +53,31 @@ BOOL PyObject_AsVARIANTRecordInfo(PyObject *ob, VARIANT *pv)
     PyRecordBuffer *owner = NULL;
+    dimNo = SafeArrayGetDim(psa);
+    if (dimNo == 0) {
+        goto exit;
```
`hr` will not be set in this error case?

geppi commented:
After reviewing the code path for the assignment of values to Record fields, I think the problem is that it was wrong from the beginning.

When I implemented the creation of _SAFEARRAY(VT_RECORD)_ from a sequence of COM Records #2317, I followed the existing code path into `PyCom_VariantFromPyObject` and changed it to fit my actual use case. Later I encountered another COM Record type which has a field defined as _SAFEARRAY(double)_ in the IDL file and tried to apply a similar change in the current PR, to also fit this use case.

Obviously IDL files in the field could contain `struct` definitions for SAFEARRAYs with other contained types, so we would need to change `PyCom_VariantFromPyObject` for the SAFEARRAYs of other types as well.
However, what would really be a problem is that an IDL file could also have a field defined as _SAFEARRAY(VARIANT)_ which would then create an unresolvable conflict for this implementation.

The culprit is the `setattro` method of `PyRecord` which ignores the type information for Record fields which would well be available.

I think `setattro` should have never taken the code path via `PyCom_VariantFromPyObject`.
Instead it should retrieve the type information for the Record field via `ITypeInfo` and call `PyCom_SAFEARRAYFromPyObject` directly using this information in the `vt` parameter. For scalar types it should call the `MakeObjToVariant` method of `PythonOleArgHelper`, again using the retrieved type information in the `m_reqdType` attribute.

I have now reverted my changes in `PyCom_VariantFromPyObject` to the state before PR #2317, so that this function keeps doing what its name says, i.e. it always creates a _VARIANT_ from PyObjects. Therefore the SAFEARRAYs it creates will continue to contain _VARIANT_ elements.

To fulfill the type requirements defined in an IDL file for the `struct` fields, the `setattro` method of `PyRecord` is changed in the way described above.

@mhammond Please let me know if you agree with this in general.

geppi commented:
@mhammond Apart from the details, are you in general OK with the revised approach for setting COM Record field values?

geppi commented:
@mhammond I'm feeling a little bit ignored. Although I don't want to push this too much, I would really like to know if you agree with the approach to retrieve the `ITypeInfo` of the Record field in `setattro` and if this properly addresses the concern you raised in [this](https://github.com/mhammond/pywin32/pull/2655#discussion_r2386320289) comment, please?

geppi commented:
Thanks.
Let me say that it's absolutely admirable that you keep this up over such a long time.

I can completely understand that larger PRs do take some time to get reviewed.
It was just that I started to worry if I had done something to get ghosted.

Please, in the future don't hesitate to just reply with a "sorry, I currently don't have the bandwidth to deal with this, please give me a ping in about one or two months".

You did this in your first comment September last year but then it went completely quiet which made me feel a little lost.
